truffle/src/main/java/org/jruby/truffle/core/array/ArrayNodes.java

```diff
@@ -911,6 +916,56 @@ protected Object fillFallback(VirtualFrame frame, DynamicObject array, Object[]
 
     }
 
+    @CoreMethod(names = "hash_internal")
+    public abstract static class HashNode extends ArrayCoreMethodNode {
+
+        @Child private ToIntNode toIntNode;
+
+        @Specialization(guards = "isNullArray(array)")
+        public long hashNull(VirtualFrame frame, DynamicObject array) {
+            final int arraySize = 0;
+            long h = Helpers.hashStart(getContext().getJRubyRuntime(), arraySize);
+            h = Helpers.murmurCombine(h, System.identityHashCode(ArrayNodes.class));
+            h = hashEnd(h);
+            return h;
+        }
+
+        @Specialization(guards = "strategy.matches(array)", limit = "ARRAY_STRATEGIES")
+        public long hash(VirtualFrame frame, DynamicObject array,
+                         @Cached("of(array)") ArrayStrategy strategy,
+                         @Cached("new()") SnippetNode snippetNode) {
```

Comment:
This could be a simple `CallDispatchHeadNode` (which can be created with `createMethodCall`).